Sources/Extraction/Filters/LocalHistogram.c

```diff
@@ -2,36 +2,34 @@
 
 #include "LocalHistogram.h"
 
-Int16Array3D LocalHistogram_Analyze(const BlockMap *blocks, const UInt8Array2D *image)
+void LocalHistogram_Analyze(const BlockMap *blocks, const UInt8Array2D *image, Int16Array3D *histogram)
 {
-    Int16Array3D histogram = Int16Array3D_Construct(blocks->blockCount.width, blocks->blockCount.height, 256);
-
     Point block;
     for (block.x = RectangleC_GetLeft(&blocks->allBlocks); block.x < RectangleC_GetRight(&blocks->allBlocks); ++block.x)
+    {
         for (block.y = RectangleC_GetBottom(&blocks->allBlocks); block.y < RectangleC_GetTop(&blocks->allBlocks); ++block.y)
         {
             RectangleC area = RectangleGrid_GetRectangleCFromPoint(&blocks->blockAreas, &block);
             for (int x = RectangleC_GetLeft(&area); x < RectangleC_GetRight(&area); ++x)
                 for (int y = RectangleC_GetBottom(&area); y < RectangleC_GetTop(&area); ++y)
-                    ++histogram.data[block.x][block.y][image->data[x][y]];
+                    ++histogram->data[block.x][block.y][image->data[x][y]];
         }
-    return histogram;
-}
-
-Int16Array3D LocalHistogram_SmoothAroundCorners(const BlockMap *blocks, const Int16Array3D *input)
-{
-    Int16Array3D output;
-
-    /* TODO: Implement LocalHistogram_SmoothAroundCorners() */
-
-    return output;
+    }
 }
 
-Int16Array3D LocalHistogram_Smooth(const BlockMap *blocks, const Int16Array3D *input)
+void LocalHistogram_SmoothAroundCorners(const Int16Array3D *input, Int16Array3D *output)
 {
-    Int16Array3D output;
-
-    /* TODO: Implement LocalHistogram_Smooth() */
-
-    return output;
+    for (int x = 0; x < input->sizeX; x++)
+    {
+        for (int y = 0; y < input->sizeY; y++)
+        {
+            for (int z = 0; z < input->sizeZ; z++)
+            {
+                output->data[x][y][z] += input->data[x][y][z];
+                output->data[x][y+1][z] += input->data[x][y][z];
```

Comment:
We'll need to check for out of range access now as output should be the same size as input